mono/mini/mini-posix.c

```diff
@@ -521,120 +504,58 @@ mono_runtime_cleanup_handlers (void)
 	free_saved_signal_handlers ();
 }
 
-#ifdef HAVE_LINUX_RTC_H
-#include <linux/rtc.h>
-#include <sys/ioctl.h>
-#include <fcntl.h>
-static int rtc_fd = -1;
+static MonoNativeThreadId sampling_thread;
+static volatile gint32 sampling_thread_running;
 
-static int
-enable_rtc_timer (gboolean enable)
+static mono_native_thread_return_t
+sampling_thread_func (void *data)
 {
-	int flags;
-	flags = fcntl (rtc_fd, F_GETFL);
-	if (flags < 0) {
-		perror ("getflags");
-		return 0;
-	}
-	if (enable)
-		flags |= FASYNC;
-	else
-		flags &= ~FASYNC;
-	if (fcntl (rtc_fd, F_SETFL, flags) == -1) {
-		perror ("setflags");
-		return 0;
+	/*
+	 * This gives us access to lock-free data structures that use hazard
+	 * pointers, e.g. the LLS (used for the thread list).
+	 */
+	mono_thread_info_register_small_id ();
+
+	MonoNativeThreadId me = mono_native_thread_id_get ();
+	guint64 us = 1000000 / mono_profiler_get_sampling_rate () - 1;
+
+	while (InterlockedRead (&sampling_thread_running)) {
+		FOREACH_THREAD_SAFE (info) {
+			if (mono_thread_info_get_tid (info) == me)
+				continue;
+
+			mono_threads_pthread_kill (info, SIGPROF);
+		} FOREACH_THREAD_SAFE_END
+
+		mono_thread_info_usleep (us);
 	}
-	return 1;
+
+	return NULL;
 }
-#endif
 
 void
 mono_runtime_shutdown_stat_profiler (void)
 {
-#ifdef HAVE_LINUX_RTC_H
-	if (rtc_fd >= 0)
-		enable_rtc_timer (FALSE);
-#endif
-}
-
-#ifdef ITIMER_PROF
-static int
-get_itimer_mode (void)
```

Comment:
So to be clear, `CLOCK_REALTIME` for the `MONO_PROFILER_STAT_MODE_REAL` case and `CLOCK_PROCESS_CPUTIME_ID` for the `MONO_PROFILER_STAT_MODE_PROCESS` case?